labrad/fasttypes.c

```diff
@@ -0,0 +1,2364 @@
+    COBJ *ft_create_cobj_from_uint32(unsigned int n);
+    COBJ *ft_create_cobj_from_float64(double x);
+    COBJ *ft_create_cobj_from_complex128(double x, double y);
+
+    COBJ *ft_create_cobj(int type, int istype);
+    
+    void ft_free_cobj(COBJ *cobj);
+    
+// functions to create a representation of the type of a COBJ
+COBJ *ft_create_cobj_type(COBJ *cobj);
+    COBJ *ft_create_list_type(COBJ *cobj);
+
+// necessarily complex function that takes an object/type and a type and makes the first more general to encompass the second. eg: int, uint returns float since this
+// is the smallest type able to represent every int and uint value. When the second argument contains placeholder values (such as v[]) it can be modified. eg: v[s], v[]
+// will not only return v[s] but modify the second argument to be v[s].
+COBJ *ft_upgrade_first_cobj_to_encompass_second(COBJ *cobj1, COBJ *cobj2);
+    COBJ *ft_upgrade_first_subtype_to_encompass_second(COBJ *cobj1, COBJ *cobj2);
+    COBJ *ft_convert_ndarray_to_list(COBJ *cobj);
+    COBJ *ft_create_cobj_from_ndarray_i(COBJ *cobj, int i);
+    int ft_size(COBJ *cobj);
+    int ft_size_subtype(COBJ *cobj);
+    
+// functions to convert a type tag into a cobj type
+COBJ *ft_create_cobj_tag_from_pystring(PyObject *tt);
+    COBJ *ft_create_cobj_tag_from_char_buf(char *buf, int *iptr, int n);
+
+// functions to initialize the global variables
+int ft_initialize();
+    int ft_system_supported();
+    void ft_detect_system_endianness();
+    int ft_set_send_endianness(PyObject *endianness);
+
+// function to print a cobj
+void ft_print_cobj(COBJ *cobj);
+    void ft_print_type(COBJ *cobj);
+    void ft_print_subtype(COBJ *cobj);
+    void ft_print_type_text(int type);
+    
+// after much preprocessing to build a clean C obj and C type, flatten
+PyObject *ft_flatten_cobj(COBJ *cobj, COBJ *cobj_type);
+    char *ft_create_type_string(COBJ *cobj, int *len_ptr);
+    char *ft_append_string(char *str1, int *len1_ptr, char *str2, int len2);
+    char *ft_append_char(char *str1, int *len1_ptr, char c);
+    int ft_write_to_buf(COBJ *cobj, char *buf);
+    void ft_appropriate_memcpy(void *dest, void *src, int len);
+    void ft_reverse_memcpy(void *dest, void *src, int len);
+
+// C equivalent of labrad.types.unflatten, called in similar manner, although type tags must be strings
+PyObject *ft_unflatten(PyObject *self, PyObject *args);
+    PyObject *ft_unflatten_partial_parse(PyObject *o);
+    PyObject *ft_unflatten_no_parse(char *s, int *si_ptr, COBJ *cobj_type);
+
+// currently immature test function
+PyObject *ft_test(PyObject *self, PyObject *args);
+
+PyObject *ft_flatten(PyObject *self, PyObject *args, PyObject *keywds) {
+    int i, n;
+    PyObject *obj, *types, *endianness, *sobj, *lobj, *tobj, *robj;
+    char *kwlist[] = {"obj", "types", "endianness", NULL};
+    COBJ *cobj, *cobj_type, *cobj_tag;
+    
+    if (!ft_gv_initialized) {
+        if (ft_initialize() == EXCEPTION_RAISED) goto exception;
```

Comment:
Yup, standard way of handling exceptions in the Python C API.
On Apr 22, 2015 4:00 PM, "pomalley" <notifications@github.com> wrote:

> In labrad/fasttypes.c
> <https://github.com/martinisgroup/pylabrad/pull/88#discussion_r28923066>:
>
> > +// C equivalent of labrad.types.unflatten, called in similar manner, although type tags must be strings
> > +PyObject *ft_unflatten(PyObject *self, PyObject *args);
> > +    PyObject *ft_unflatten_partial_parse(PyObject *o);
> > +    PyObject *ft_unflatten_no_parse(char *s, int *si_ptr, COBJ *cobj_type);
> > +
> > +// currently immature test function
> > +PyObject *ft_test(PyObject *self, PyObject *args);
> > +
> > +PyObject *ft_flatten(PyObject *self, PyObject *args, PyObject *keywds) {
> > +    int i, n;
> > +    PyObject *obj, *types, *endianness, *sobj, *lobj, *tobj, *robj;
> > +    char *kwlist[] = {"obj", "types", "endianness", NULL};
> > +    COBJ *cobj, *cobj_type, *cobj_tag;
> > +
> > +    if (!ft_gv_initialized) {
> > +        if (ft_initialize() == EXCEPTION_RAISED) goto exception;
>
> goto!!
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/martinisgroup/pylabrad/pull/88/files#r28923066>.
>